In [ ]:
!python -m pip install -U python-woc pandas matplotlib
# please clear the output of this cell in your notebook before checking it in

In [1]:
from woc.remote import WocMapsRemote
from tqdm import tqdm
import pandas as pd
# creates the client
woc = WocMapsRemote( base_url="https://worldofcode.org/api/")
# if you got an API key
# woc = WocMapsRemote( base_url="https://worldofcode.org/api/", api_key="woc-XXXXXX-YYYYYY" )

# Now for each of the ten projects you were assigned get commits, e.g.
projects = ['cybergis_job-supervisor',
            'libfive_libfive',
            'hpi-information-systems_metanome',
            'nasa_smcpy',
            'brownhci_drafty',
            'andergray_probabilityboundsanalysis.jl',
            'computationalmodelling_fidimag',
            'wsshin_maxwellfdfd',
            'pytorch_fbgemm',
            'stardist_stardist']
list_df_commits = []
for prj in projects:
  # Convert GitHub repository names to WoC V2412 project names
  #prj = prj.lower().replace('/','_',2)
  commits = woc.get_values('p2c', prj)
  df = pd.DataFrame(commits, columns=["sha1"])
  df['project'] = prj
  list_df_commits.append(df)
df = pd.concat(list_df_commits)
#perhaps save the list (if no errors) so you do not need to retrieve them again
df.to_csv('df_commits.csv')
df.head(1)

,sha1,project
0,0075ff18028044c525d5d6dfc17926081b3a889b,cybergis_job-supervisor


In [2]:
# If the number of commits its not very large, you can try to get them all at the same time,
# The max batch size is 10
import time
# let us first split df['sha1'] in chunks
chunks = [df['sha1'][x:x+10] for x in range(0, len(df), 10)]
commit_data = []

for chunk in tqdm(chunks): # iterate over the commits
  # res, err = woc.show_content_many('commit',chunk.to_list())

  # commit.tch returns the same data but faster
  res, err = woc.get_values_many('commit.tch',chunk.to_list())
  res = {k: v[0] for k, v in res.items()}  # this conversion is necessary because of the internal implementation

  # to walk around the rate limit
  time.sleep(1)

  if err: # check for errors
    print('Got Errors', err)

  for commit_sha, commit in res.items():
    # flatten commit objects
    commit_data.append({
          'commit': commit_sha,
          'tree': commit[0],
          'parent': list(commit[1]),
          'author': commit[2][0],
          'author_time': int(commit[2][1]),
          'author_tz': commit[2][2],
          'committer': commit[3][0],
          'committer_time': int(commit[3][1]),
          'committer_tz': commit[3][2],
          'message': commit[4],
    })

df_commit_data = pd.DataFrame(commit_data)
df_commit_data = df_commit_data.merge(df, left_on='commit', right_on='sha1')
df_commit_data.to_csv('df_commit_data.csv', index=False)
df_commit_data.head(2)

  5%|▌         | 158/3145 [02:53<53:53,  1.08s/it] 

Got Errors {'16d36be32a171fdd0b5f5b3bcf967099af846fc1': 'Key 16d36be32a171fdd0b5f5b3bcf967099af846fc1 not found in /da5_fast/All.sha1c/commit_22.tch'}


 18%|█▊        | 579/3145 [10:31<49:37,  1.16s/it]

Got Errors {'394cb53e1f389dca891822832823060ba9f9805a': 'Key 394cb53e1f389dca891822832823060ba9f9805a not found in /da5_fast/All.sha1c/commit_57.tch'}


 20%|█▉        | 619/3145 [11:16<45:47,  1.09s/it]

Got Errors {'5c0879eea224cfd76aba88aa87b2e8fb4b23f7b2': 'Key 5c0879eea224cfd76aba88aa87b2e8fb4b23f7b2 not found in /da5_fast/All.sha1c/commit_92.tch'}


 22%|██▏       | 679/3145 [12:22<43:54,  1.07s/it]

Got Errors {'91250ad59816d42274fe09625dff3232d8aade73': 'Key 91250ad59816d42274fe09625dff3232d8aade73 not found in /da5_fast/All.sha1c/commit_17.tch'}


 22%|██▏       | 700/3145 [12:44<43:38,  1.07s/it]

Got Errors {'a2f26aba3395a7ccf081bd14800251e5c85aef89': 'Key a2f26aba3395a7ccf081bd14800251e5c85aef89 not found in /da5_fast/All.sha1c/commit_34.tch'}


 24%|██▍       | 755/3145 [13:43<43:03,  1.08s/it]

Got Errors {'d4e6bc940424641967489f1767a9eedbc2b10bb0': 'Key d4e6bc940424641967489f1767a9eedbc2b10bb0 not found in /da5_fast/All.sha1c/commit_84.tch'}


 44%|████▍     | 1396/3145 [25:21<31:57,  1.10s/it]

Got Errors {'fd84da8808af839216e9abe383018d00172a7c58': 'Key fd84da8808af839216e9abe383018d00172a7c58 not found in /da5_fast/All.sha1c/commit_125.tch'}


 46%|████▋     | 1460/3145 [26:30<30:34,  1.09s/it]

Got Errors {'0421bf5a83213149eca7d85b0f201a8d1c8ef88e': 'Key 0421bf5a83213149eca7d85b0f201a8d1c8ef88e not found in /da5_fast/All.sha1c/commit_4.tch'}


 49%|████▉     | 1547/3145 [28:05<29:09,  1.09s/it]

Got Errors {'11bf6de3a63dcce08a4c036c10300dcab475015a': 'Key 11bf6de3a63dcce08a4c036c10300dcab475015a not found in /da5_fast/All.sha1c/commit_17.tch'}


 53%|█████▎    | 1668/3145 [30:18<27:05,  1.10s/it]

Got Errors {'25afbddd3e1279345040ec7d72c9fd6cbd02d3f2': 'Key 25afbddd3e1279345040ec7d72c9fd6cbd02d3f2 not found in /da5_fast/All.sha1c/commit_37.tch'}


 55%|█████▍    | 1725/3145 [31:20<25:44,  1.09s/it]

Got Errors {'2e95cbe94369badb6a062c89a471e3f4d2cb44bf': 'Key 2e95cbe94369badb6a062c89a471e3f4d2cb44bf not found in /da5_fast/All.sha1c/commit_46.tch'}


 61%|██████    | 1918/3145 [34:54<22:50,  1.12s/it]

Got Errors {'4e773ea229c4183a362e5ee6550244d757106379': 'Key 4e773ea229c4183a362e5ee6550244d757106379 not found in /da5_fast/All.sha1c/commit_78.tch'}


 66%|██████▌   | 2083/3145 [37:55<19:23,  1.10s/it]

Got Errors {'68f031e2b961874c4a1676987a7239c497a53579': 'Key 68f031e2b961874c4a1676987a7239c497a53579 not found in /da5_fast/All.sha1c/commit_104.tch'}


 67%|██████▋   | 2101/3145 [38:15<19:15,  1.11s/it]

Got Errors {'6bcc6412bfbbd71f2fcf447731b927bc12e9f0db': 'Key 6bcc6412bfbbd71f2fcf447731b927bc12e9f0db not found in /da5_fast/All.sha1c/commit_107.tch'}


 67%|██████▋   | 2113/3145 [38:28<18:46,  1.09s/it]

Got Errors {'6d95a72f6465f90d02820f5b91a4c9f0bb3141d5': 'Key 6d95a72f6465f90d02820f5b91a4c9f0bb3141d5 not found in /da5_fast/All.sha1c/commit_109.tch'}


 74%|███████▎  | 2319/3145 [42:13<15:00,  1.09s/it]

Got Errors {'8f37582319587d2e60e4c88271575613f26f6717': 'Key 8f37582319587d2e60e4c88271575613f26f6717 not found in /da5_fast/All.sha1c/commit_15.tch'}


 86%|████████▌ | 2700/3145 [49:08<08:03,  1.09s/it]

Got Errors {'cd4134c49d5de7d37021e92878d7f212dd027ce1': 'Key cd4134c49d5de7d37021e92878d7f212dd027ce1 not found in /da5_fast/All.sha1c/commit_77.tch'}


 86%|████████▌ | 2703/3145 [49:12<08:04,  1.10s/it]

Got Errors {'cd98d41ef52acbd70117274dafe59ac4e6c85b30': 'Key cd98d41ef52acbd70117274dafe59ac4e6c85b30 not found in /da5_fast/All.sha1c/commit_77.tch'}


 88%|████████▊ | 2761/3145 [50:15<07:08,  1.12s/it]

Got Errors {'d6f5a367e290a356532aa60ec6d29b059d7b8ccf': 'Key d6f5a367e290a356532aa60ec6d29b059d7b8ccf not found in /da5_fast/All.sha1c/commit_86.tch'}


 88%|████████▊ | 2782/3145 [50:38<06:39,  1.10s/it]

Got Errors {'dabf57702a91860224567c6aea06e3d5ac47adf3': 'Key dabf57702a91860224567c6aea06e3d5ac47adf3 not found in /da5_fast/All.sha1c/commit_90.tch'}


 93%|█████████▎| 2921/3145 [53:12<04:05,  1.09s/it]

Got Errors {'f147b0c10ab3fcc3417c6cafbb08724b0647e4c9': 'Key f147b0c10ab3fcc3417c6cafbb08724b0647e4c9 not found in /da5_fast/All.sha1c/commit_113.tch'}


100%|██████████| 3145/3145 [57:17<00:00,  1.09s/it]


,commit,tree,parent,author,author_time,author_tz,committer,committer_time,committer_tz,message,sha1,project
0,0075ff18028044c525d5d6dfc17926081b3a889b,4fb25a249e469352cf8e85bc8f988f1ca1d68c30,[2495fdcf6ed4d2ea6692b9cdd7792f9851d2400f],Alexander Michels <michels9@illinois.edu>,1750783227,-0500,GitHub <noreply@github.com>,1750783227,-0500,speed up git fetch and updated containers to u...,0075ff18028044c525d5d6dfc17926081b3a889b,cybergis_job-supervisor
1,010a5b360d94b6c978c08452b9c41e24e9d46dd4,c685b1b3547e512d9dab7d6d9792a03bd748a1db,[d284286f67d287539a1f15806f7664431f7e3908],Ian Zhang <yianzhang14@gmail.com>,1740535063,-0600,Ian Zhang <yianzhang14@gmail.com>,1740535063,-0600,more syncing\n,010a5b360d94b6c978c08452b9c41e24e9d46dd4,cybergis_job-supervisor


In [3]:
# what does the first record looks like?
# commit  parents
v = commit_data[0]
# commit tree, [parent{s}], [author, unixtime, tz ], [ committer, unixtime, tz ], Commit_message ]
print(v)

{'commit': '0075ff18028044c525d5d6dfc17926081b3a889b', 'tree': '4fb25a249e469352cf8e85bc8f988f1ca1d68c30', 'parent': ['2495fdcf6ed4d2ea6692b9cdd7792f9851d2400f'], 'author': 'Alexander Michels <michels9@illinois.edu>', 'author_time': 1750783227, 'author_tz': '-0500', 'committer': 'GitHub <noreply@github.com>', 'committer_time': 1750783227, 'committer_tz': '-0500', 'message': 'speed up git fetch and updated containers to use new path on anvil (#156)\n\nCo-authored-by: mitkotak <zyli2004@gmail.com>'}


# Make sure you check in to your fork not just the notebook but also the csv files!!!

# Don't forget to add requested data from github and this notebook
### For each of the 10 projects go to their github repo and get the number of stars, number of forks, and the last commit date
### Report (in your notebook) the number of commits, the number of authors, and max and min time for each project based on WoC commits and also add the info you obtained from github

In [8]:
netid = 'hcrettol'

# Save all the commits to one csv file
out = df_commit_data[['project', 'commit', 'author', 'author_time', 'message']]
out.to_csv(netid + '_project_summary.csv', sep=';', index=False, header=False)
print('Saved', len(out), 'commits')

# Count commits and authors for each project, and find the first and last commit
stats = out.groupby('project').agg(
    ncommits=('commit', 'count'),
    nauthors=('author', 'nunique'),
    min_time=('author_time', 'min'),
    max_time=('author_time', 'max'))

# Turn the timestamps into normal dates
stats['from'] = pd.to_datetime(stats['min_time'], unit='s').dt.strftime('%Y-%m-%d')
stats['to'] = pd.to_datetime(stats['max_time'], unit='s').dt.strftime('%Y-%m-%d')
print(stats)

Saved 31421 commits
                                        ncommits  nauthors    min_time  \
project                                                                  
andergray_probabilityboundsanalysis.jl       406         8  1566923172   
brownhci_drafty                             2490        21  1454018185   
computationalmodelling_fidimag              2007        45  1354627521   
cybergis_job-supervisor                     1198        27  1594862452   
hpi-information-systems_metanome            2834        54  1363423355   
libfive_libfive                             3975       114  1448854231   
nasa_smcpy                                  1064        28  1536353601   
pytorch_fbgemm                             15766       852  1539384493   
stardist_stardist                           1312        41  1530047944   
wsshin_maxwellfdfd                           369         4  1351309510   

                                          max_time        from          to  
project       

| Project | Commits | Authors | From | To | Stars | Forks | Last GH Commit |
|---|---|---|---|---|---|---|---|
| andergray_probabilityboundsanalysis.jl | 406 | 8 | 2019-08-27 | 2025-03-10 | 26 | 2 | 2025-01-13 |
| brownhci_drafty | 2490 | 21 | 2016-01-28 | 2025-08-01 | 9 | 4 | 2025-05-29 |
| computationalmodelling_fidimag | 2007 | 45 | 2012-12-04 | 2025-10-17 | 68 | 27 | 2026-09-28 |
| cybergis_job-supervisor | 1198 | 27 | 2020-07-16 | 2025-10-31 | 12 | 5 | 2025-06-24 |
| hpi-information-systems_metanome | 2834 | 54 | 2013-03-16 | 2025-08-25 | 194 | 68 | 2025-06-05 |
| libfive_libfive | 3975 | 114 | 2015-11-30 | 2025-08-19 | 1700 | 177 | 2025-11-12 |
| nasa_smcpy | 1064 | 28 | 2018-09-07 | 2025-09-16 | 137 | 37 | 2026-07-24 |
| pytorch_fbgemm | 15766 | 852 | 2018-10-12 | 2025-11-06 | 1600 | 790 | 2026-09-28 |
| stardist_stardist | 1312 | 41 | 2018-06-26 | 2025-07-31 | 1300 | 283 | 2026-02-14 |
| wsshin_maxwellfdfd | 369 | 4 | 2012-10-27 | 2025-04-16 | 123 | 51 | 2025-04-15 |